File: backend/internal/models/workflow.go
package models

import "time"

// Workflow represents a complete workflow with steps and variables
type Workflow struct {
	ID          string     `json:"id"`
	Name        string     `json:"name"`
	Description string     `json:"description"`
	Category    string     `json:"category"`
	Variables   []Variable `json:"variables"`
	Steps       []Step     `json:"steps"`
	CreatedAt   time.Time  `json:"created_at"`
	UpdatedAt   time.Time  `json:"updated_at"`
}

// Variable represents a workflow variable (global or local)
type Variable struct {
	Name         string   `json:"name"`
	Description  string   `json:"description"`
	Type         string   `json:"type"` // string, number, select, boolean
	DefaultValue string   `json:"default_value,omitempty"`
	Options      []string `json:"options,omitempty"` // For select type
	IsGlobal     bool     `json:"is_global"`         // Global or template-specific
	Required     bool     `json:"required"`
}

// Step represents a single step in a workflow
type Step struct {
	ID         string            `json:"id"`
	Name       string            `json:"name"`
	Type       string            `json:"type"` // command, workflow_ref, conditional
	Order      int               `json:"order"`
	Content    string            `json:"content"`              // Command or workflow_id
	Variables  map[string]string `json:"variables,omitempty"`  // Variable mappings
	Conditions []Condition       `json:"conditions,omitempty"` // Conditional logic
	OnSuccess  *StepAction       `json:"on_success,omitempty"`
	OnFailure  *StepAction       `json:"on_failure,omitempty"`
}

// Condition represents a conditional check on step output
type Condition struct {
	Type     string     `json:"type"`  // contains, equals, starts_with, ends_with, regex, exit_code
	Value    string     `json:"value"` // Value to compare
	Operator string     `json:"operator,omitempty"` // AND, OR (for multiple conditions)
	Action   StepAction `json:"action"`
}

// StepAction defines what to do based on condition result
type StepAction struct {
	Type   string `json:"type"`   // continue, stop, jump_to, execute_step
	Target string `json:"target"` // Step ID or workflow ID
}

// WorkflowExecution tracks the execution state
type WorkflowExecution struct {
	ID         string            `json:"id"`
	WorkflowID string            `json:"workflow_id"`
	Status     string            `json:"status"` // pending, running, completed, failed, cancelled
	Variables  map[string]string `json:"variables"`
	Logs       []ExecutionLog    `json:"logs"`
	StartTime  string            `json:"start_time"`
	EndTime    string            `json:"end_time,omitempty"`
}

// ExecutionLog represents a single log entry during execution
type ExecutionLog struct {
	Timestamp string `json:"timestamp"`
	StepID    string `json:"step_id,omitempty"`
	Level     string `json:"level"` // info, warning, error
	Message   string `json:"message"`
	Output    string `json:"output,omitempty"`
}

// GlobalVariable represents a reusable global variable
type GlobalVariable struct {
	Name        string    `json:"name"`
	Value       string    `json:"value"`
	Description string    `json:"description"`
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`
}

File: frontend/lib/models/workflow.dart
class Workflow {
  final String id;
  final String name;
  final String description;
  final String category;
  final List<VariableDefinition> variables;
  final List<WorkflowStep> steps;

  Workflow({
    required this.id,
    required this.name,
    required this.description,
    required this.category,
    required this.variables,
    required this.steps,
  });

  factory Workflow.fromJson(Map<String, dynamic> json) {
    return Workflow(
      id: json['id'] ?? '',
      name: json['name'] ?? '',
      description: json['description'] ?? '',
      category: json['category'] ?? '',
      variables: (json['variables'] as List<dynamic>?)
              ?.map((e) => VariableDefinition.fromJson(e))
              .toList() ??
          [],
      steps: (json['steps'] as List<dynamic>?)
              ?.map((e) => WorkflowStep.fromJson(e))
              .toList() ??
          [],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'description': description,
      'category': category,
      'variables': variables.map((e) => e.toJson()).toList(),
      'steps': steps.map((e) => e.toJson()).toList(),
    };
  }
}

class VariableDefinition {
  final String name;
  final String description;
  final String type; // string, number, boolean, select
  final String? defaultValue;
  final List<String>? options;

  VariableDefinition({
    required this.name,
    required this.description,
    required this.type,
    this.defaultValue,
    this.options,
  });

  factory VariableDefinition.fromJson(Map<String, dynamic> json) {
    return VariableDefinition(
      name: json['name'] ?? '',
      description: json['description'] ?? '',
      type: json['type'] ?? 'string',
      defaultValue: json['default_value'],
      options: (json['options'] as List<dynamic>?)?.map((e) => e.toString()).toList(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'description': description,
      'type': type,
      'default_value': defaultValue,
      'options': options,
    };
  }
}

class WorkflowStep {
  final String id;
  final String name;
  final String type; // command, workflow_ref
  final String content;
  final Map<String, String> variables;
  final StepCondition? condition;

  WorkflowStep({
    required this.id,
    required this.name,
    required this.type,
    required this.content,
    this.variables = const {},
    this.condition,
  });

  factory WorkflowStep.fromJson(Map<String, dynamic> json) {
    return WorkflowStep(
      id: json['id'] ?? '',
      name: json['name'] ?? '',
      type: json['type'] ?? 'command',
      content: json['content'] ?? '',
      variables: Map<String, String>.from(json['variables'] ?? {}),
      condition: json['condition'] != null ? StepCondition.fromJson(json['condition']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'type': type,
      'content': content,
      'variables': variables,
      'condition': condition?.toJson(),
    };
  }
}

class StepCondition {
  final String type; // contains, equals, regex, exit_code
  final String value;
  final String action; // continue, stop, jump_to_step
  final String? target;

  StepCondition({
    required this.type,
    required this.value,
    required this.action,
    this.target,
  });

  factory StepCondition.fromJson(Map<String, dynamic> json) {
    return StepCondition(
      type: json['type'] ?? 'contains',
      value: json['value'] ?? '',
      action: json['action'] ?? 'continue',
      target: json['target'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'type': type,
      'value': value,
      'action': action,
      'target': target,
    };
  }
}

class WorkflowExecution {
  final String id;
  final String workflowId;
  final String status;
  final List<String> logs;
  final Map<String, String> variables;
  final String startTime;
  final String? endTime;

  WorkflowExecution({
    required this.id,
    required this.workflowId,
    required this.status,
    required this.logs,
    required this.variables,
    required this.startTime,
    this.endTime,
  });

  factory WorkflowExecution.fromJson(Map<String, dynamic> json) {
    return WorkflowExecution(
      id: json['id'] ?? '',
      workflowId: json['workflow_id'] ?? '',
      status: json['status'] ?? 'pending',
      logs: (json['logs'] as List<dynamic>?)?.map((e) => e.toString()).toList() ?? [],
      variables: Map<String, String>.from(json['variables'] ?? {}),
      startTime: json['start_time'] ?? '',
      endTime: json['end_time'],
    );
  }
}

class GlobalVariable {
  final String name;
  final String value;
  final String description;
  final DateTime createdAt;
  final DateTime updatedAt;

  GlobalVariable({
    required this.name,
    required this.value,
    required this.description,
    required this.createdAt,
    required this.updatedAt,
  });

  factory GlobalVariable.fromJson(Map<String, dynamic> json) {
    return GlobalVariable(
      name: json['name'] ?? '',
      value: json['value'] ?? '',
      description: json['description'] ?? '',
      createdAt: DateTime.tryParse(json['created_at'] ?? '') ?? DateTime.now(),
      updatedAt: DateTime.tryParse(json['updated_at'] ?? '') ?? DateTime.now(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'value': value,
      'description': description,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
    };
  }
}
